synapseutils/copy.py

```diff
@@ -194,24 +195,31 @@ def _copyRecursive(syn, entity, destinationId, mapping=None, skipCopyAnnotations
 
-    if isinstance(ent, Project):
-        if not isinstance(syn.get(destinationId), Project):
-            raise ValueError("You must give a destinationId of a new project to copy projects")
-        copiedId = destinationId
-        entities = syn.getChildren(entity)
-        for i in entities:
-            mapping = _copyRecursive(syn, i['id'], destinationId, mapping=mapping,
-                                     skipCopyAnnotations=skipCopyAnnotations, **kwargs)
-    elif isinstance(ent, Folder):
-        copiedId = _copyFolder(syn, ent.id, destinationId, mapping=mapping, skipCopyAnnotations=skipCopyAnnotations,
-                               **kwargs)
-    elif isinstance(ent, File) and "file" not in excludeTypes:
-        copiedId = _copyFile(syn, ent.id, destinationId, version=version, updateExisting=updateExisting,
-                             setProvenance=setProvenance, skipCopyAnnotations=skipCopyAnnotations)
-    elif isinstance(ent, Link) and "link" not in excludeTypes:
-        copiedId = _copyLink(syn, ent.id, destinationId, updateExisting=updateExisting)
-    elif isinstance(ent, Schema) and "table" not in excludeTypes:
-        copiedId = _copyTable(syn, ent.id, destinationId, updateExisting=updateExisting)
+    profile_username = syn.getUserProfile()['userName']
+    permissions = syn.getPermissions(ent, profile_username)
+
+    access_requirements = syn.restGET('/entity/%s/accessRequirement' % ent.id)
+
+    #Do not copy anything with access requirements
+    if access_requirements['totalNumberOfResults'] == 0:
+        if isinstance(ent, Project):
+            if not isinstance(syn.get(destinationId), Project):
+                raise ValueError("You must give a destinationId of a new project to copy projects")
+            copiedId = destinationId
+            entities = syn.getChildren(entity)
+            for i in entities:
+                mapping = _copyRecursive(syn, i['id'], destinationId, mapping=mapping,
+                                         skipCopyAnnotations=skipCopyAnnotations, **kwargs)
+        elif isinstance(ent, Folder):
+            copiedId = _copyFolder(syn, ent.id, destinationId, mapping=mapping, skipCopyAnnotations=skipCopyAnnotations,
+                                   **kwargs)
+        elif isinstance(ent, File) and "file" not in excludeTypes and "DOWNLOAD" in permissions:
+            copiedId = _copyFile(syn, ent.id, destinationId, version=version, updateExisting=updateExisting,
+                                 setProvenance=setProvenance, skipCopyAnnotations=skipCopyAnnotations)
+        elif isinstance(ent, Link) and "link" not in excludeTypes:
+            copiedId = _copyLink(syn, ent.id, destinationId, updateExisting=updateExisting)
+        elif isinstance(ent, Schema) and "table" not in excludeTypes and "DOWNLOAD" in permissions:
+            copiedId = _copyTable(syn, ent.id, destinationId, updateExisting=updateExisting)
```

Comment:
separate PR for this

Reply:
for other PR, probably can just check DOWNLOAD in permissions once for any entity type